src/main/java/info/skyblond/nekohit/neo/contract/WCAContract.java

```diff
@@ -65,33 +62,40 @@
     @DisplayName("Refund")
     private static Event4Args<Hash160, String, Integer, Integer> onRefund;
 
+    @DisplayName("CancelWCA")
+    private static Event1Arg<String> onCancelWCA;
+
 
     @OnNEP17Payment
     public static void onPayment(Hash160 from, int amount, Object data) throws Exception {
-        require(CAT_TOKEN_HASH == Runtime.getCallingScriptHash(), "Only Cat Token can invoke this function.");
-        require(amount > 0, "Transfer amount must be a positive number.");
+        require(CAT_TOKEN_HASH == Runtime.getCallingScriptHash(), Messages.INVALID_CALLER);
+        require(amount > 0, Messages.INVALID_AMOUNT);
         String identifier = (String) data;
         WCABasicInfo basicInfo = getWCABasicInfo(identifier);
-        require(basicInfo != null, "Identifier not found.");
+        require(basicInfo != null, Messages.RECORD_NOT_FOUND);
         List<WCAMilestone> milestones = getWCAMilestones(identifier);
-        require(milestones != null, "Identifier not found.");
+        require(milestones != null, Messages.BROKEN_RECORD);
 
         if (basicInfo.owner.equals(from)) {
             // owner paying stake
-            require(!basicInfo.paid, "You can't pay a paid WCA.");
-            require(!checkIfReadyToFinish(milestones), "You can't pay a expired WCA.");
-            require(basicInfo.getTotalStake() == amount, "Amount not correct");
-            // unpaid before, not finished(expired), amount is correct
-            basicInfo.paid = true;
-            wcaBasicInfoMap.put(identifier, StdLib.serialize(basicInfo));
+            require(basicInfo.status == 0, Messages.INVALID_STATUS_ALLOW_PENDING);
+            require(basicInfo.getTotalStake() == amount, Messages.INCORRECT_AMOUNT);
+            // unpaid before, amount is correct, set to OPEN
+            basicInfo.status = 1;
+            // update status in case the threshold milestone is expired
+            basicInfo.updateStatus(milestones);
+            updateWCABasicInfo(identifier, basicInfo);
             onPayWCA.fire(from, identifier, amount);
         } else {
-            // buyer want to buy a WCA
-            throwIfNotAvailableToBuy(basicInfo, milestones);
+            // buyer want to buy a WCA, update status first
+            basicInfo.updateStatus(milestones);
+            require(basicInfo.status == 1 || basicInfo.status == 2, Messages.INVALID_STATUS_ALLOW_OPEN_AND_ACTIVE);
+            require(!checkIfReadyToFinish(milestones), Messages.INVALID_STATUS_READY_TO_FINISH);
             WCABuyerInfo buyerInfo = getWCABuyerInfo(identifier);
-            require(buyerInfo != null, "Buyer info not found.");
+            require(buyerInfo != null, Messages.BROKEN_RECORD);
             buyerInfo.recordPurchase(from, amount);
-            wcaBuyerInfoMap.put(identifier, StdLib.serialize(buyerInfo));
+            updateWCABasicInfo(identifier, basicInfo);
```

Comment:
this updates every time, even when there are no changes. A little bit better would be to check if there is even a change but then there will be a buyer who has to pay a bit more GAS when there is actually a change so this is also not great.
Adhoc I have no better idea but I would add a check if something changed before.

Reply:
`updateWCABasicInfo` has an if statement internally and only updates once if the condition is met. After that, the new status should fail the if statement. But by updating the BasicInfo, the content will be rewritten and waste some GAS if the BasicInfo is not changed. 

I'm wondering if there is some way to extract those fields to a new POJO and make it as small as possible since we need to rewrite the status anyway. Maybe split the static fields into one POJO, like description, milestones. They will only be write-once and won't be changed. We can store the proof of work in another POJO, which will be rewritten on finishing the milestones. For some fields like this, a normal integer is 4byte, aka 4 English char, which shouldn't cost too much GAS.

Reply:
hm, couldn't we add the status to the BuyerInfo? we are updating it anyways

Reply:
If we have a lot of buyers, then we will repeat a large list of buyers' addresses and their purchase amount. For now, I have no idea how to solve this.

Reply:
#24